Gems/Terrain/Code/Tests/TerrainSystemTest.cpp

```diff
@@ -1037,4 +1056,74 @@ namespace UnitTest
 
         terrainSystem->ProcessSurfacePointsFromRegion(testRegionBox, stepSize, perPositionCallback, AzFramework::Terrain::TerrainDataRequests::Sampler::EXACT);
     }
+
+    TEST_F(TerrainSystemTest, TerrainProcessAsyncCancellation)
+    {
+        // Tests cancellation of the asynchronous terrain API.
+
+        const AZ::Aabb spawnerBox = AZ::Aabb::CreateFromMinMaxValues(-10.0f, -10.0f, -5.0f, 10.0f, 10.0f, 15.0f);
+        auto entity = CreateAndActivateMockTerrainLayerSpawner(
+            spawnerBox,
+            [](AZ::Vector3& position, bool& terrainExists)
+            {
+                // Our generated height will be X + Y.
+                position.SetZ(position.GetX() + position.GetY());
+                terrainExists = true;
+            });
+
+        // Create and activate the terrain system with our testing defaults for world bounds, and a query resolution at 1 meter intervals.
+        auto terrainSystem = CreateAndActivateTerrainSystem();
+
+        // Generate some input positions (doesn't really matter how many as we'll cancel while the first is being processed).
+        AZStd::vector<AZ::Vector3> inPositions;
+        for (int i = 0; i < 64; ++i)
+        {
+            inPositions.push_back({1.0f, 1.0f, 1.0f});
+        }
+
+        // Setup the per position callback so that we can cancel the entire request when it is first invoked.
+        AZStd::atomic_bool asyncRequestStarted = false;
+        AZStd::atomic_bool asyncRequestCancelled = false;
```

Comment:
This should use binary_semaphore or something more "yielding" than a busy loop on an atomic_bool.

Reply:
This test is irrelevant if we're going back to the original version which doesn't allow cancellation after the job has already started.

Reply:
I re-introduced the test, but changed it to use a binary_semaphore, good suggestion.